source/validate_types.cpp

```diff
@@ -335,36 +359,54 @@ SpvMemoryModel ValidationState_t::getMemoryModel() const {
   return memory_model_;
 }
 
-Functions::Functions(ValidationState_t& module)
-    : module_(module), in_function_(false), in_block_(false) {}
-
-bool Functions::in_function_body() const { return in_function_; }
-
-bool Functions::in_block() const { return in_block_; }
+Function::Function(uint32_t id, uint32_t result_type_id,
+                   SpvFunctionControlMask function_control,
+                   uint32_t function_type_id, ValidationState_t& module)
+    : module_(module),
+      id_(id),
+      function_type_id_(function_type_id),
+      result_type_id_(result_type_id),
+      function_control_(function_control),
+      declaration_type_(FunctionDecl::kFunctionDeclUnknown),
+      blocks_(),
+      current_block_(nullptr),
+      cfg_constructs_(),
+      variable_ids_(),
+      parameter_ids_() {}
+
+bool Function::in_block() const { return static_cast<bool>(current_block_); }
+
+bool Function::IsFirstBlock(uint32_t id) const {
+  return *get_first_block() == id;
+}
 
-spv_result_t Functions::RegisterFunction(uint32_t id, uint32_t ret_type_id,
-                                         uint32_t function_control,
-                                         uint32_t function_type_id) {
+spv_result_t ValidationState_t::RegisterFunction(
+    uint32_t id, uint32_t ret_type_id, SpvFunctionControlMask function_control,
+    uint32_t function_type_id) {
+  assert(in_function_ == false &&
+         "Function instructions can not be declared in a function");
   assert(in_function_ == false &&
          "Function instructions can not be declared in a function");
   in_function_ = true;
-  id_.emplace_back(id);
-  type_id_.emplace_back(function_type_id);
-  declaration_type_.emplace_back(FunctionDecl::kFunctionDeclUnknown);
-  block_ids_.emplace_back();
-  variable_ids_.emplace_back();
-  parameter_ids_.emplace_back();
+  module_functions_.emplace_back(id, ret_type_id, function_control,
+                                 function_type_id, *this);
 
   // TODO(umar): validate function type and type_id
-  (void)ret_type_id;
-  (void)function_control;
 
   return SPV_SUCCESS;
 }
 
-spv_result_t Functions::RegisterFunctionParameter(uint32_t id,
-                                                  uint32_t type_id) {
-  assert(in_function_ == true &&
+spv_result_t ValidationState_t::RegisterFunctionEnd() {
+  assert(in_function_body() == true &&
+         "Function end can only be called in functions");
+  assert(in_block() == false && "Function end cannot be called inside a block");
```

Comment:
On second thought, this could go into a separate PR that systematically addresses it everywhere.